src/hotspot/share/runtime/mrswMutex.hpp

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_RUNTIME_MRSWMUTEX_HPP
+#define SHARE_RUNTIME_MRSWMUTEX_HPP
+
+#include "javaThread.hpp"
+#include "memory/allocation.hpp"
+#include "runtime/atomic.hpp"
+#include "runtime/mutex.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/macros.hpp"
+
+class NoTransition;
+
+// Multi-reader single-writer lock implementation.
+// If a transition needs to occur when the reader or writer lock is taken (in
+// case the thread blocks), then specialize on the ReaderTransition and/or
+// WriterTransition parameters. This lock is unfair, high contention of readers
+// may starve some of them.
+// Writers take precedence, blocking new readers from entering and allowing
+// current readers to proceed. The consequences are undefined if one or more
+// writer threads attempt to enter their critical region when another writer
+// thread already is attempting to do so or currently is in its critical region.
```

Comment:
>Are you saying that this is truly restricted to only a single thread being allowed to use this in write mode, rather than just the normal one-writer-at-a-time? 

The comment was wrong, it's just the regular 1-writer-at-a-time.